imxrt-hal/src/can/mod.rs

```diff
@@ -0,0 +1,1114 @@
+use crate::ral;
+
+use core::convert::Infallible;
+use core::marker::PhantomData;
+
+/// Error that indicates that an incoming message has been lost due to buffer overrun.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+#[cfg_attr(feature = "unstable-defmt", derive(defmt::Format))]
+pub struct OverrunError {
+    _priv: (),
+}
+
+/// Unclocked Can modules
+///
+/// The `Unclocked` struct represents the unconfigured Can peripherals.
+/// Once clocked, you'll have the ability to build Can peripherals from the
+/// compatible processor pins.
+pub struct Unclocked {
+    pub(crate) can1: ral::can::Instance,
+    pub(crate) can2: ral::can::Instance,
+}
+
+impl Unclocked {
+    /// Enable clocks to all Can modules, returning a builder for the two Can modules.
+    pub fn clock(
+        self,
+        handle: &mut ccm::Handle,
+        clock_select: ccm::can::ClockSelect,
+        divider: ccm::can::PrescalarSelect,
+    ) -> (Builder<U1>, Builder<U2>) {
+        let (ccm, _) = handle.raw();
+        // First, disable clocks
+        ral::modify_reg!(
+            ral::ccm,
+            ccm,
+            CCGR1,
+            CG0: 0,
+            CG1: 0,
+            CG2: 0,
+            CG3: 0
+        );
+
+        let clk_sel = match clock_select {
+            ccm::can::ClockSelect::Pll2 => ral::ccm::CSCMR2::CAN_CLK_SEL::RW::CAN_CLK_SEL_1,
+        };
+
+        // Select clock, and commit prescalar
+        ral::modify_reg!(
+            ral::ccm,
+            ccm,
+            CSCMR2,
+            CAN_CLK_PODF: ral::ccm::CSCMR2::CAN_CLK_PODF::RW::DIVIDE_1,
+            CAN_CLK_SEL: clk_sel
+        );
+
+        // Enable clocks
+        ral::modify_reg!(
+            ral::ccm,
+            ccm,
+            CCGR1,
+            CG0: 0b11,
+            CG1: 0b11,
+            CG2: 0b11,
+            CG3: 0b11
+        );
+
+        let source_clock = ccm::Frequency::from(clock_select) / ccm::Divider::from(divider);
+        (
+            Builder::new(source_clock, self.can1),
+            Builder::new(source_clock, self.can2),
+        )
+    }
+}
+
+/// A Can builder that can build a Can peripheral
+pub struct Builder<M> {
+    _module: PhantomData<M>,
+    reg: ral::can::Instance,
+    source_clock: ccm::Frequency,
+}
+
+impl<M> Builder<M>
+where
+    M: Unsigned,
+{
+    fn new(source_clock: ccm::Frequency, reg: ral::can::Instance) -> Self {
+        Builder {
+            _module: PhantomData,
+            reg,
+            source_clock,
+        }
+    }
+
+    /// Builds a Can peripheral. The return
+    /// is a configured FlexCan peripheral running at 24MHz.
+    pub fn build(self) -> CAN<M> {
+        CAN::new(self.source_clock, self.reg)
+    }
+}
+
+/// A Can master
+///
+pub struct CAN<M> {
+    reg: ral::can::Instance,
+    _module: PhantomData<M>,
+    source_clock: ccm::Frequency,
+    _mailbox_reader_index: u8,
+}
+
+const CAN1_ADDR: u32 = 0x401d0000;
+const CAN2_ADDR: u32 = 0x401d4000;
```

Comment:
Future-proofing consideration in light of #121: CAN1 and CAN2 are at different addresses on (at least) the 1170. It looks like these are only needed for the `base_address()` implementation.

- If we could remove `base_address()`, then we could also remove these two constants.
- If we need to keep `base_address()`, let's acquire the base address from the `imxrt-ral` instance.

    ```rust
    pub fn base_address(&self) -> u32 {
        let addr: *const ral::can::RegisterBlock = &*self.reg;
        addr as u32
    }
    ```

    The `imxrt-ral` instance is (expected to be) correct no matter the target MCU.

Reply:
The base address is needed to access FlexCAN mailboxes, which are not part of the RAL imxrt-ral crate at this point. I removed the constants and modified the `base_address()` as above.